File: ResourceManager/ViewModels/Books/EbookPopupViewModel.cs
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Options;
using Microsoft.Win32;
using NetCore.Mvvm.Abstractions;
using NetCore.Mvvm.Commands;
using NetCore.Mvvm.Controls;
using NetCore.Mvvm.ViewModels;
using NetCore.Validators;
using ResourceManager.Models;
using ResourceManager.Services.Abstractions;
using ResourceManager.Settings;
using System;
using System.IO;
using System.Threading.Tasks;
using System.Windows.Input;

namespace ResourceManager.ViewModels
{
    public class EbookPopupViewModel : PopupViewModelBase
    {
        private readonly IDataServiceManager _dataServiceManager;
        private int ID;
        private bool coverIsChanged = false;
        private bool fileIsChanged = false;
        private bool isEdit;
        private Book oldBook;

        public EbookPopupViewModel(IDataServiceManager dataServiceManager)
        {
            _dataServiceManager = dataServiceManager;
            isEdit = false;

            AddValidationRule(() => Title, TextValidator.Instance.Required, TextValidator.Instance.ErrorMessage("Required", nameof(Title)));
            AddValidationRule(() => Author, TextValidator.Instance.Required, TextValidator.Instance.ErrorMessage("Required", nameof(Author)));
            AddValidationRule(() => FilePath, TextValidator.Instance.Required, TextValidator.Instance.ErrorMessage("Required", nameof(FilePath)));
            AddValidationRule(() => Publisher, TextValidator.Instance.Required, TextValidator.Instance.ErrorMessage("Required", nameof(Publisher)));
            
            SaveCommand = new AsyncRelayCommand(OnSaveAsync, CanSave);
            BrowseFileCommand = new RelayCommand(OnBrowseFile);
            BrowseCoverCommand = new RelayCommand(OnBrowseCover);
        }
      
        private void OnBrowseCover()
        {
            var dlg = new OpenFileDialog();
            dlg.Title = "Select cover file";
            dlg.Filter = App.Settings.EbookSettings.CoverFileExtensions;
            dlg.CheckFileExists = true;
            var res = dlg.ShowDialog();
            if(res!=null && res.Equals(true))
            {
                CoverPath = dlg.FileName;
                coverIsChanged = true;
            }
        }

        private void OnBrowseFile()
        {
            var dlg = new OpenFileDialog();
            dlg.Title = "Select ebook file";
            dlg.Filter = App.Settings.EbookSettings.SupportExtensions;
            dlg.CheckFileExists = true;
            var res = dlg.ShowDialog();
            if (res != null && res.Equals(true))
            {
                FilePath = dlg.FileName;
                fileIsChanged = true;
                if (!isEdit)
                {
                    Title = string.IsNullOrWhiteSpace(Title) ? Path.GetFileNameWithoutExtension(dlg.FileName) : Title;
                    Author = string.IsNullOrWhiteSpace(Author) ? "-" : string.Empty;
                    Publisher = string.IsNullOrWhiteSpace(Publisher) ? "-" : string.Empty;
                }
            }
        }

        private bool CanSave()
        {
            return !HasErrors;
        }
        public void SetBook(Book book)
        {
            isEdit = true;
            oldBook = book;
            ID = book.ID;
            Title= book.Title;
            Author= book.Author;
            Publisher= book.Publisher;
            FilePath = book.FilePath;
            CoverPath = book.CoverPath;
            Abstraction = book.Abstraction;
        }
        private async Task OnSaveAsync()
        {
            ValidateAllRules();

            if (!HasErrors)
            {
                string generatedFilename = string.Concat(Title.Split(Path.GetInvalidFileNameChars(), StringSplitOptions.RemoveEmptyEntries));
                if (generatedFilename.Length > 250) generatedFilename = generatedFilename.Substring(0, 250);
                try
                {
                    if (isEdit)
                    {
                        string destFilename = generatedFilename + Path.GetExtension(FilePath);
                        string destPath = Path.Combine(App.Settings.EbookSettings.FolderPath, destFilename);
                        if (fileIsChanged)
                        {
                            if (App.Settings.EbookSettings.DeleteSourceFile)
                                File.Move(FilePath, destPath, true);
                            else
                                File.Copy(FilePath, destPath, true);
                        }
                        else
                        {
                            if (destFilename != oldBook.Filename)
                            {
                                File.Move(oldBook.FilePath, destPath);
                            }
                        }
                        string coverFilename = generatedFilename + Path.GetExtension(CoverPath);
                        string coverPath = Path.Combine(App.Settings.EbookSettings.CoverPath, coverFilename);
                        if (coverIsChanged)
                        {
                            if (!string.IsNullOrEmpty(CoverPath))
                            {
                                if (App.Settings.EbookSettings.DeleteSourceFile)
                                    File.Move(CoverPath, coverPath, true);
                                else
                                    File.Copy(CoverPath, coverPath, true);
                            }
                        }
                        else
                        {
                            if (!string.IsNullOrEmpty(CoverPath))
                            {
                                if (coverFilename != oldBook.Cover)
                                {
                                    File.Move(CoverPath, coverPath);
                                }
                            }
                            else
                            {
                                coverFilename = string.Empty;
                            }
                        }
                        var book = new Book()
                        {
                            Title = Title,
                            Author = Author,
                            Publisher = Publisher,
                            Filename = destFilename,
                            Cover = coverFilename,
                            Abstraction = Abstraction
                        };
                        await _dataServiceManager.BookDataService.Update(ID, book);

                        Result = PopupResult.OK;

                    }
                    else
                    {
                        string destFilename = generatedFilename + Path.GetExtension(FilePath);
                        string destPath = Path.Combine(App.Settings.EbookSettings.FolderPath, destFilename);
                        if (App.Settings.EbookSettings.DeleteSourceFile)
                            File.Move(FilePath, destPath, true);
                        else
                            File.Copy(FilePath, destPath, true);
                        string coverFilename = string.Empty;
                        string coverPath = string.Empty;
                        if (!string.IsNullOrEmpty(CoverPath))
                        {
                            coverFilename = generatedFilename + Path.GetExtension(CoverPath);
                            coverPath = Path.Combine(App.Settings.EbookSettings.CoverPath, coverFilename);
                            if (App.Settings.EbookSettings.DeleteSourceFile)
                                File.Move(CoverPath, coverPath, true);
                            else
                                File.Copy(CoverPath, coverPath, true);
                        }
                        var book = new Book()
                        {
                            Title = Title,
                            Author = Author,
                            Publisher = Publisher,
                            Filename = destFilename,
                            Cover = coverFilename,
                            Abstraction = Abstraction
                        };
                        await _dataServiceManager.BookDataService.Create(book);

                        Result = PopupResult.OK;


                    }
                }
                catch (Exception e)
                {
                    windowManager.ShowMessageBox(e.Message, System.Windows.MessageBoxButton.OK, System.Windows.MessageBoxImage.Error);
                }
                OnClose();
            }
        }

        public string Title
        {
            get => GetProperty<string>();
            set => SetProperty(value);
        }
        public string Author
        {
            get => GetProperty<string>();
            set => SetProperty(value);
        }
        public string Publisher
        {
            get => GetProperty<string>();
            set => SetProperty(value);
        }
        public string FilePath
        {
            get => GetProperty<string>();
            set => SetProperty(value);
        }
        public string CoverPath
        {
            get => GetProperty<string>();
            set => SetProperty(value);
        }
        public string Abstraction
        {
            get => GetProperty<string>();
            set => SetProperty(value);
        }
        public ICommand SaveCommand { get; }
        public ICommand BrowseFileCommand { get; }
        public ICommand BrowseCoverCommand { get; }
    }
}

File: ResourceManager/ViewModels/Books/EbookPopupViewModel.vb
Imports Microsoft.Win32
Imports VBNetCore.Mvvm.Abstractions
Imports VBNetCore.Mvvm.Commands
Imports VBNetCore.Mvvm.Controls
Imports VBNetCore.Mvvm.ViewModels
Imports VBNetCore.Validators
Imports ResourceManager.Models
Imports ResourceManager.Services.Abstractions
Imports System
Imports System.IO
Imports System.Threading.Tasks
Imports System.Windows.Input

Namespace ResourceManager.ViewModels
    Public Class EbookPopupViewModel
        Inherits PopupViewModelBase
        Private ReadOnly _dataServiceManager As IDataServiceManager
        Private ID As Integer
        Private coverIsChanged As Boolean = False
        Private fileIsChanged As Boolean = False
        Private isEdit As Boolean
        Private oldBook As Book

        Public Sub New(ByVal dataServiceManager As IDataServiceManager)
            _dataServiceManager = dataServiceManager
            isEdit = False

            AddValidationRule(Function() Title, Function(x) TextValidator.Instance.Required(x), TextValidator.Instance.ErrorMessage("Required", NameOf(EbookPopupViewModel.Title)))
            AddValidationRule(Function() Author, Function(x) TextValidator.Instance.Required(x), TextValidator.Instance.ErrorMessage("Required", NameOf(EbookPopupViewModel.Author)))
            AddValidationRule(Function() FilePath, Function(x) TextValidator.Instance.Required(x), TextValidator.Instance.ErrorMessage("Required", NameOf(EbookPopupViewModel.FilePath)))
            AddValidationRule(Function() Publisher, Function(x) TextValidator.Instance.Required(x), TextValidator.Instance.ErrorMessage("Required", NameOf(EbookPopupViewModel.Publisher)))

            SaveCommand = New AsyncRelayCommand(AddressOf OnSaveAsync, AddressOf CanSave)
            BrowseFileCommand = New RelayCommand(AddressOf OnBrowseFile)
            BrowseCoverCommand = New RelayCommand(AddressOf OnBrowseCover)
        End Sub

        Private Sub OnBrowseCover()
            Dim dlg = New OpenFileDialog With {
                .Title = "Select cover file",
                .Filter = Application.Settings.EbookSettings.CoverFileExtensions,
                .CheckFileExists = True
            }
            Dim res = dlg.ShowDialog()
            If res IsNot Nothing AndAlso res.Equals(True) Then
                CoverPath = dlg.FileName
                coverIsChanged = True
            End If
        End Sub

        Private Sub OnBrowseFile()
            Dim dlg = New OpenFileDialog()
            dlg.Title = "Select ebook file"
            dlg.Filter = Application.Settings.EbookSettings.SupportExtensions
            dlg.CheckFileExists = True
            Dim res = dlg.ShowDialog()
            If res IsNot Nothing AndAlso res.Equals(True) Then
                FilePath = dlg.FileName
                fileIsChanged = True
                If Not isEdit Then
                    Title = If(String.IsNullOrWhiteSpace(Title), Path.GetFileNameWithoutExtension(dlg.FileName), Title)
                    Author = If(String.IsNullOrWhiteSpace(Author), "-", String.Empty)
                    Publisher = If(String.IsNullOrWhiteSpace(Publisher), "-", String.Empty)
                End If
            End If
        End Sub

        Private Function CanSave() As Boolean
            Return Not HasErrors
        End Function
        Public Sub SetBook(ByVal book As Book)
            isEdit = True
            oldBook = book
            ID = book.ID
            Title = book.Title
            Author = book.Author
            Publisher = book.Publisher
            FilePath = book.FilePath
            CoverPath = book.CoverPath
            Abstraction = book.Abstraction
        End Sub
        Private Async Function OnSaveAsync() As Task
            ValidateAllRules()

            If Not HasErrors Then
                Dim generatedFilename As String = FileValidator.Instance.Filter(Title)
                If generatedFilename.Length > 250 Then generatedFilename = generatedFilename.Substring(0, 250)
                Try
                    If isEdit Then
                        Dim destFilename = generatedFilename & Path.GetExtension(FilePath)
                        Dim destPath = Path.Combine(Application.Settings.EbookSettings.FolderPath, destFilename)
                        If fileIsChanged Then
                            If Application.Settings.EbookSettings.DeleteSourceFile Then
                                File.Move(FilePath, destPath, True)
                            Else
                                File.Copy(FilePath, destPath, True)
                            End If
                        Else
                            If Not Equals(destFilename, oldBook.Filename) Then
                                File.Move(oldBook.FilePath, destPath)
                            End If
                        End If
                        Dim coverFilename = generatedFilename & Path.GetExtension(Me.CoverPath)
                        Dim coverPath = Path.Combine(Application.Settings.EbookSettings.CoverPath, coverFilename)
                        If coverIsChanged Then
                            If Not String.IsNullOrEmpty(Me.CoverPath) Then
                                If Application.Settings.EbookSettings.DeleteSourceFile Then
                                    File.Move(Me.CoverPath, coverPath, True)
                                Else
                                    File.Copy(Me.CoverPath, coverPath, True)
                                End If
                            End If
                        Else
                            If Not String.IsNullOrEmpty(Me.CoverPath) Then
                                If Not Equals(coverFilename, oldBook.Cover) Then
                                    File.Move(Me.CoverPath, coverPath)
                                End If
                            Else
                                coverFilename = String.Empty
                            End If
                        End If
                        Dim book = New Book() With {
    .Title = Title,
    .Author = Author,
    .Publisher = Publisher,
    .Filename = destFilename,
    .Cover = coverFilename,
    .Abstraction = Abstraction
}
                        Await _dataServiceManager.BookDataService.Update(ID, book)

                        Result = PopupResult.OK
                    Else
                        Dim destFilename = generatedFilename & Path.GetExtension(FilePath)
                        Dim destPath = Path.Combine(Application.Settings.EbookSettings.FolderPath, destFilename)
                        If Application.Settings.EbookSettings.DeleteSourceFile Then
                            File.Move(FilePath, destPath, True)
                        Else
                            File.Copy(FilePath, destPath, True)
                        End If
                        Dim coverFilename = String.Empty
                        Dim coverPath = String.Empty
                        If Not String.IsNullOrEmpty(Me.CoverPath) Then
                            coverFilename = generatedFilename & Path.GetExtension(Me.CoverPath)
                            coverPath = Path.Combine(Application.Settings.EbookSettings.CoverPath, coverFilename)
                            If Application.Settings.EbookSettings.DeleteSourceFile Then
                                File.Move(Me.CoverPath, coverPath, True)
                            Else
                                File.Copy(Me.CoverPath, coverPath, True)
                            End If
                        End If
                        Dim book = New Book() With {
    .Title = Title,
    .Author = Author,
    .Publisher = Publisher,
    .Filename = destFilename,
    .Cover = coverFilename,
    .Abstraction = Abstraction
}
                        Await _dataServiceManager.BookDataService.Create(book)

                        Result = PopupResult.OK


                    End If
                Catch e As Exception
                    windowManager.ShowMessageBox(e.Message, System.Windows.MessageBoxButton.OK, System.Windows.MessageBoxImage.Error)
                End Try
                OnClose()
            End If
        End Function

        Public Property Title As String
            Get
                Return GetProperty(Of String)()
            End Get
            Set(ByVal value As String)
                SetProperty(value)
            End Set
        End Property
        Public Property Author As String
            Get
                Return GetProperty(Of String)()
            End Get
            Set(ByVal value As String)
                SetProperty(value)
            End Set
        End Property
        Public Property Publisher As String
            Get
                Return GetProperty(Of String)()
            End Get
            Set(ByVal value As String)
                SetProperty(value)
            End Set
        End Property
        Public Property FilePath As String
            Get
                Return GetProperty(Of String)()
            End Get
            Set(ByVal value As String)
                SetProperty(value)
            End Set
        End Property
        Public Property CoverPath As String
            Get
                Return GetProperty(Of String)()
            End Get
            Set(ByVal value As String)
                SetProperty(value)
            End Set
        End Property
        Public Property Abstraction As String
            Get
                Return GetProperty(Of String)()
            End Get
            Set(ByVal value As String)
                SetProperty(value)
            End Set
        End Property
        Public ReadOnly Property SaveCommand As ICommand
        Public ReadOnly Property BrowseFileCommand As ICommand
        Public ReadOnly Property BrowseCoverCommand As ICommand
    End Class
End Namespace
